testplan/common/utils/match.py

```diff
@@ -139,23 +333,43 @@ def match(
         if isinstance(regex.pattern, str):
             read_mode = "r"
 
-        with open(self.log_path, read_mode) as log:
-            log.seek(self.position)
+        try:
+            log_handle = self.strategy.open_logfile_in_position(
+                position=self.position, mode=read_mode
+            )
 
             while True:
-                if timeout > 0 and time.time() > end_time:
+                if (timeout > 0) and (time.time() > end_time):
                     break
-                line = log.readline()
+
+                line = log_handle.readline()
                 if line:
                     match = regex.match(line)
                     if match:
                         break
-                elif timeout > 0:
-                    time.sleep(LOG_MATCHER_INTERVAL)
                 else:
-                    break
 
-            self.position = log.tell()
+                    previous_log_handle = log_handle
+
+                    log_handle, is_new_file = self.strategy.get_next_file(
```

Comment:
Can we absort the get_next_file thing into strategy?

I think the strategy layer shall serve as a virtual and continuous log file, it provides readline() method that LogMatcher can call without knowing we are stepping into another rotated log segment.

Maybe Strategy is not the right name - i'm really talking about an abstraction of the rotating log.

Reply:
We can call it LogStream or LogSource than, what the interface would be

- `readline() -> Union[str|None]`
- `position() -> Position`  Position is an opaque that can be just used for comparision, and for seek
- `seek(p: Position)` 
- `sof() -> Position`
- `eof() -> Position`
- `compare(position1: Position, position2: Position) -> int` classic compare meaning
- `close()` maybe for optimalization from matcher to close the stream when it does not need it for a while.

I like that we have a single component touching the file handles. 

The question is what is the lifetime of the object, 
 - Is it a member of the Matcher, then is it hold a file handle to the logfile at actual position? When does it close it, ? This would remove the need to store the position in the Matcher
 - Is it recreated and seeked for every match operation as the file is opened today.

Reply:
Combined with the internal comment and some offline discussion, if a file name pattern in regex + inode create time for putting things in order will suffice for most of the use cases with log rotation -- I don't think we need to introduce new public class here for user to extend, but would like to make the out-of-box usage really simple and straight-forward and trim down the lines of code we need to maintain.

I will suggest a usage like following:

log_matcher = LogMatcher(log_path="path/to/log/file", rotation_pattern="engine1.*log")

If rotation_pattern is specified, we can create a LogStream inside LogMatcher. I think the LogStream instance can be long-living in LogMatcher, but then the underlying log file, maybe open/close every time match() is called (this is what we do today with non-rotating logfile).

Reply:
I do not like the bit where we leave out the potential for extensibility configuring LogMatcher with something LogStream could make this much more useful. Boxing this into pattern and directory would make this again a breaking change when a new need is coming.

Reply:
Do you have any potential use case in mind, that would go beyond the file name pattern in regex?

Do you prefer something like this:

log_matcher = LogMatcher(log_path="path/to/log/file", log_rotation_strategy=RegexStrategy(pattern="engine1.*log")

where RegexStrategy is provided out- of-box, and will only be used to identify the group of rotated logs.

Reply:
I think maybe a log_path="path/to/log/engine*log" might be a good enough heuristic for now. We can introduce a strategy parameter in future if different requirement come up, then we don't break anything.